software/controller/test/hal/test_spi_chain/spi_chain_test.cpp

```diff
@@ -0,0 +1,345 @@
+/* Copyright 2022, RespiraWorks
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "gtest_main.h"
+#include "mock_spi_chain.h"
+
+using namespace SPI;
+
+// Helper function that enqueues a SPI and updates the expected spi flow (if provided,
+// along with the number of slaves in the chain)
+template <size_t N>
+void EnqueueRequestAndTestResult(DaisyChain<N> *chain, Request &request, size_t slave,
+                                 bool expected_result, uint8_t *expected_flow = nullptr,
+                                 size_t num_slaves = 0) {
+  EXPECT_EQ(chain->SendRequest(request, slave), expected_result);
+  if (expected_result) {
+    EXPECT_FALSE(*(request.processed));
+
+    if (expected_flow) {
+      for (size_t byte = 0; byte < request.length; byte++) {
+        expected_flow[byte * num_slaves] = request.command[byte];
+      }
+    }
+  }
+}
+
+// Helper function that checks that the daisy chain state matches the expected one.
+template <size_t N>
+void CheckDaisyState(MockChain<N> *chain, uint8_t *expected, size_t num_slaves) {
+  uint8_t chain_state[num_slaves] = {0};
+  chain->GetChainState(chain_state, num_slaves);
+  for (size_t slave = 0; slave < num_slaves; slave++) {
+    SCOPED_TRACE("slave=" + std::to_string(slave));
+    EXPECT_EQ(chain_state[slave], expected[slave]);
+  }
+}
+
+TEST(SpiDaisyChain, RequestQueue) {
```

Comment:
Maybe we could have smaller tests? And they should be declarative, like "InvalidCommandProvokesErrorInterrupt" etc..

Reply:
We could indeed have a few small test cases which could be more declarative, but I would still need a couple of big ones for dequeuing, and those ones would pretty much repeat all the simple tests, because they require a lot of setting up.
What I am doing here is using all the steps I need to go through when setting up the dequeuing tests to test simpler stuff along the way, resulting in less runtime, as I am not doing the simple tests twice.

And even with seemingly simple tests, it seems inefficient to fill up a request queue to be able to trigger the `queue_count_[slave] >= MaxQueueLength` condition and not do anything else in that test case, when I have many other seemingly simple test cases which require a full queue (can I still queue to slave 1 while slave 0 has a full queue? can I queue again after having emptied that full queue? ...)